### ISSUE TITLE
Should escape literal </script> 

### ISSUE THREAD

**Issue by wch** [jeroenooms/jsonlite#163]
When JSON is embedded in a web page, it could have content like this:

```
<script>
   {"x":"<script></script>"}
</script>
```

The first `</script>` (on the second line) closes the opening `<script>` tag. This is a security risk if the string contains something like this:

```
<script>
   {"x":"</script><script>evil()</script>"}
</script>
```

See https://benalpert.com/2012/08/03/preventing-xss-json.html for more info.

This can also result in other problems. For example: https://github.com/rstudio/rmarkdown/issues/943#issuecomment-275160192

I believe the right thing to do is to escape `</` strings as `<\/`.

**Comment by jeroenooms** [comment]
@wch Unfortunately the change in default escaping of `/` has broken quite a few packages. For example this [unit test in httr](https://github.com/hadley/httr/blob/master/tests/testthat/test-oauth-server-side.R#L20) now fails because it compares a base64 encoded json blob, contains urls, against a literal string.

I'm going to change this to only escape `/` when it is preceded by a `<`. That solves the problem right?

**Comment by wch** [comment]
Yup, that should solve the problem. I have a partial implementation of this, but ran into the `prettify` and `minify` problem mentioned here: https://github.com/jeroenooms/jsonlite/pull/164#issuecomment-275736274

Here's the diff:

```diff
$ git diff
diff --git a/src/escape_chars.c b/src/escape_chars.c
index bab1f6e..06e0d32 100644
--- a/src/escape_chars.c
+++ b/src/escape_chars.c
@@ -1,5 +1,6 @@
 #include <Rinternals.h>
 #include <stdlib.h>
+#include <stdbool.h>
 
 /*
 Fast escaping of character vectors (Winston Chang)
@@ -14,6 +15,9 @@ SEXP C_escape_chars_one(SEXP x) {
 
   // Count the number of matches
   int matches = 0;
+  // Keep track of whether previous character '<'. This is because we escape
+  // "</" to "<\/", but don't escape other instances of "/".
+  bool prev_open_angle = false;
   do {
     switch(*cur) {
       case '\\':
@@ -23,8 +27,18 @@ SEXP C_escape_chars_one(SEXP x) {
       case '\t':
       case '\b':
       case '\f':
-      case '/':
         matches++;
+        prev_open_angle = false;
+        break;
+      case '/':
+        // Only need to count '/' if following '<'
+        if (prev_open_angle)
+          matches++;
+        prev_open_angle = false;
+        break;
+      case '<':
+        prev_open_angle = true;
+        break;
     }
     cur++;
   } while(*cur); //breaks on '\0'
@@ -40,50 +54,69 @@ SEXP C_escape_chars_one(SEXP x) {
   char * outcur = newstr;
   *outcur++ = '"';
 
+  prev_open_angle = false;
+
   do {
     switch(*cur) {
       case '\\':
         *outcur++ = '\\';
         *outcur = '\\';
+        prev_open_angle = false;
         break;
       case '"':
         *outcur++ = '\\';
         *outcur = '"';
+        prev_open_angle = false;
         break;
       case '\n':
         *outcur++ = '\\';
         *outcur = 'n';
+        prev_open_angle = false;
         break;
       case '\r':
         *outcur++ = '\\';
         *outcur = 'r';
+        prev_open_angle = false;
         break;
       case '\t':
         *outcur++ = '\\';
         *outcur = 't';
+        prev_open_angle = false;
         break;
       case '\b':
         *outcur++ = '\\';
         *outcur = 'b';
+        prev_open_angle = false;
         break;
       case '\f':
         *outcur++ = '\\';
         *outcur = 'f';
+        prev_open_angle = false;
         break;
       case '/':
-        *outcur++ = '\\';
+        // Escape '/' only if previous char was '<'
+        if (prev_open_angle)
+          *outcur++ = '\\';
         *outcur = '/';
+        prev_open_angle = false;
+        break;
+
+      case '<':
+        *outcur = *cur;
+        prev_open_angle = true;
         break;
 
       //this should never happen: edge case of embedded null in string
       case '\0':
         *outcur++ = '"';
         *outcur = '\0';
+        prev_open_angle = false;
         break;
 
       //simply copy char from input
       default:
         *outcur = *cur;
+        prev_open_angle = false;
     }
 
     //increment input and output cursors to next character

```

**Comment by jeroenooms** [comment]
Thanks, I had already implemented it... Can you test if it works now? 

As you mention, yajl cannot preserve the solidus, so prettify/minify will either always or never escape it. I decided that `minify()` should probably not escape the solidus to achieve the most minified encoding. People specifically use `minify()` to normalize the JSON, so I think in this case it should do what other implementations do. All implementations I've found (including JavaScript default to not escaping the solidus).

**Comment by wch** [comment]
Looks good to me.

**Comment by jeroenooms** [comment]
For future reference: I ended up hacking the `libyajl` code to mimic the same behavior in `minify()` / `prettify()`, i.e. escape solidus if and only-if it is preceded by a `<` in the string.

**Comment by jeroenooms** [comment]
@wch this has landed in v1.3 on CRAN now.

**Comment by wch** [comment]
Excellent, thanks!
